_pages/dev/ssr/server-side-rendering-error-handling.md

```diff
@@ -0,0 +1,422 @@
+
+## Multi error handlers
+Together with SSR Error Handling, Spartacus brings possibility to provide multiple error handlers that can react on the error caught during server-side rendering. The error handlers are provided in the `CxErrorHandler` via multi-provided token.
+Default multi-provided error handlers are:
+- `LoggingErrorHandler` - logs the error using `LoggerService`;
+- `PropagatingToServerErrorHandler` - passes the error to the `CxCommonEngine` thanks to `PROPAGATE_ERROR_TO_SERVER` injection token.
+
+To create a custom multi-provided error handler, you need to implement and interface `MultiErrorHandler` and provide the handler in a module.
+The following is an example of how to crate a multi-provided error handler:
+```ts
+@Injectable({
+  providedIn: 'root',
+})
+export class MyMultiErrorHandler implements MultiErrorHandler {
+  handleError(error: unknown): void {
+    /* custom error handling logic */
+  }
+}
+```
+```ts
+@NgModule({
+  providers: [
+    {
+      provide: MULTI_ERROR_HANDLER,
+      useClass: MyMultiErrorHandler,
+      multi: true,
+    },
+  ],
+})
+export class MyModule {}
+```
+
+**Note:** All multi-provided error handlers are executed. The order of execution is same as the order they are provided. 
+
+
+## Using default ExpressJS error handlers
+
+To handle errors in ExpressJS that were caught during server-side rendering, Spartacus provides a set of default error handlers that can be used in the `server.ts` file:
+
+```ts
+/* ... */
+export function app(): express.Express {
+  const server = express();
+  const distFolder = join(process.cwd(), 'dist/ng-15-to-17-ssr-eh/browser');
+  const indexHtml = existsSync(join(distFolder, 'index.original.html'))
+    ? 'index.original.html'
+    : 'index.html';
+  const indexHtmlContent = readFileSync(join(distFolder, indexHtml), 'utf8');
+
+  /* ... */
+
+  server.use(defaultExpressErrorHandlers(indexHtmlContent));
+
+  return server;
+}
+```
+
+The `defaultExpressErrorHandlers` function takes the `documentContent` as an argument, which is the content of the application's `index.html` file. The function returns an error handler which drives `Fallback to CSR` strategy and sends the `documentContent` with the proper status code based on the error type. By default, the function returns `404` status code for `CmsPageNotFoundOutboundHttpError` and `500` status code for other errors.
+```ts
+//express-error-handlers.ts
+export const defaultExpressErrorHandlers =
+  (documentContent: string): ErrorRequestHandler =>
+  (err, _req, res, _next) => {
+    if (!res.headersSent) {
+      res.set('Cache-Control', 'no-store');
+      const statusCode =
+        err instanceof CmsPageNotFoundOutboundHttpError
+          ? HttpResponseStatus.NOT_FOUND //404
+          : HttpResponseStatus.INTERNAL_SERVER_ERROR; //500
+      res.status(statusCode).send(documentContent);
+    }
+  };
+```
```

Comment:
Instead of copy-pasting code that might get outdated soon, let's put here just a live link:
https://github.com/SAP/spartacus/blob/develop/core-libs/setup/ssr/error-handling/express-error-handlers/express-error-handlers.ts